osquery/logger/plugins/aws_log_forwarder.h

```diff
@@ -0,0 +1,310 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <vector>
+
+#include <osquery/core.h>
+#include <osquery/dispatcher.h>
+#include <osquery/logger.h>
+
+#include "osquery/logger/plugins/buffered.h"
+#include "osquery/utils/aws_util.h"
+
+namespace osquery {
+template <typename RecordType,
+          typename ClientType,
+          typename OutcomeType,
+          typename ResultType>
+class AwsLogForwarder : public BufferedLogForwarder {
+ public:
+  using Outcome = OutcomeType;
+  using Client = ClientType;
+  using Record = RecordType;
+  using Batch = std::vector<RecordType>;
+  using BatchList = std::vector<Batch>;
+  using Result = ResultType;
+
+ public:
+  AwsLogForwarder(const std::string& name, size_t log_period, size_t max_lines)
+      : BufferedLogForwarder(name, std::chrono::seconds(log_period), max_lines),
+        name_(name) {}
+
+  /// Common plugin initialization
+  Status setUp() override {
+    Status s = BufferedLogForwarder::setUp();
+    if (!s.ok()) {
+      return s;
+    }
+
+    s = makeAWSClient<Client>(client_);
+    if (!s.ok()) {
+      return s;
+    }
+
+    return internalSetup();
+  }
+
+ private:
+  /// Dumps the given batch to the error log
+  void dumpBatchToErrorLog(const Batch& batch) const {
+    std::stringstream error_output;
+
+    error_output << name_
+                 << " logger: Failed to write the following records:\n";
+    dumpBatch(error_output, batch);
+
+    LOG(ERROR) << error_output.str();
+  }
+
+  /// Dumps the discarded records to the error log
+  void dumpDiscardedRecordsToErrorLog(
+      const std::vector<std::string>& discarded_records) const {
+    if (discarded_records.empty()) {
+      return;
+    }
+
+    std::stringstream output;
+    output << name_ << ": The following log records have been discarded "
+                       "because they were too big:\n";
+
+    for (const auto& record : discarded_records) {
+      output << record << "\n";
+    }
+
+    LOG(ERROR) << output.str();
+  }
+
+  /// Dumps the specified batch to the given stream
+  std::ostream& dumpBatch(std::ostream& stream, const Batch& batch) const {
```

Comment:
Looks good, but I'm concerned about this.

This sets a new precedent for dumping multiline logs. I promised to be less strict on the logging so let's see how the AWS stakeholders feel. If we can get some practical testing of debug builds we can iterate later.

@mimeframe, let me know your thoughts.

Reply:
@alessandrogario - can you explain at a high-level what this means?

Reply:
The log lines are grouped together into one or more batches (depending on the size of the log and the protocol limits). Once a batch fails to be sent to destination (maximum number of retries reached) this method dumps the unsent log lines to LOG(ERROR), so that they are not lost forever.